module/Photo/src/Photo/Service/Admin.php

```diff
@@ -0,0 +1,279 @@
+{
+    /**
+     * Generates CFS paths
+     *
+     * @param string $path The path of the photo to generate the path for
+     *
+     * @return string the path at which the photo should be saved
+     */
+    public function generateStoragePath($path)
+    {
+        $config = $this->getConfig();
+        $hash = sha1_file($path);
+        /**
+         * the hash is split to obtain a path
+         * like 92/cfceb39d57d914ed8b14d0e37643de0797ae56.jpg
+         */
+        $directory = substr($hash, 0, 2);
+        if (!file_exists($config['upload_dir'] . '/' . $directory)) {
+            mkdir($config['upload_dir'] . '/' . $directory);
+        }
+        $parts = explode('.', $path);
+        $fileType = end($parts);
+        $storagePath = $directory . '/' . substr($hash, 2) . '.' . strtolower($fileType);
+
+        return $storagePath;
+    }
+
+    /**
+     * Move the uploaded photo to the storage and store it in the database.
+     * All upload actions should use this function to prevent "ghost" files
+     * or database entries
+     *
+     * @param string $path the temporary path of the uploaded photo
+     * @param \Photo\Model\Album $targetAlbum the album to save the photo in
+     * @param boolean $move whether to move the photo instead of copying it
+     *
+     * @return \Photo\Model\Photo
+     */
+    public function storeUploadedPhoto($path, $targetAlbum, $move = false)
+    {
+        $config = $this->getConfig();
+        $storagePath = $this->generateStoragePath($path);
+        //check if photo exists already in the database
+        $photo = $this->getPhotoMapper()->getPhotoByData($storagePath, $targetAlbum);
+        //if the returned object is null, then the photo doesn't exist
+        if (is_null($photo)) {
+            $photo = new PhotoModel();
+            $photo->setAlbum($targetAlbum);
+            $photo = $this->getMetadataService()->populateMetaData($photo, $path);
+            $photo->setPath($storagePath);
+
+            //Create and set the storage paths for thumbnails.
+            //Currently, the maximum sizes of the old website are used. These
+            //values are dependant on the design.
+            $photo->setLargeThumbPath($this->createThumbnail(
+                $path,
+                $config['large_thumb_size']['width'],
+                $config['large_thumb_size']['height']
+            ));
+            $photo->setSmallThumbPath($this->createThumbnail(
+                $path,
+                $config['small_thumb_size']['width'],
+                $config['small_thumb_size']['height']
+            ));
+            $mapper = $this->getPhotoMapper();
+            /**
+             * TODO: optionally we could use a transactional query here to make it
+             * completely fail-safe in case something goes wrong when moving the
+             * photo in the storeUploadedPhoto function. However it's very unlikely
+             * anything will go wrong when moving the photo.
```

Comment:
Inderdaad een goed idee als dat eindelijk eens gaat gebeuren. (dit stond al een jaar zo in de master volgens mij)